### ISSUE TITLE
TensorRT 7.1 produce incorrect results when batch size >1

### ISSUE THREAD

**Issue by RizhaoCai** [NVIDIA/TensorRT#842]
## Environment

**TensorRT Version**:  7.1.3.0
**GPU Type**: Jetson TX2 (Jetpack 4.4)
**Nvidia Driver Version**: 
**CUDA Version**: 10.1
**CUDNN Version**: 
**Operating System + Version**: (LTS from Jetpack 4.4)
**Python Version (if applicable)**: 3.6.9
**PyTorch Version (if applicable)**: 1.6.0

## Relevant Files
|----rfb.onnx
|----trt_utils.py
|----produce_bug.py
|____________________
These files can be downloaded from [google drive](https://drive.google.com/drive/folders/1ZeLJjV9gbyEB3OwXnyx5ucvaGvFiLzRk?usp=sharing)

I reproduce produce_bug.py as below

```python
import os
import sys
import pickle
import argparse
import numpy as np
from pprint import pprint
import pdb
import torch
import torch.backends.cudnn as cudnn
cudnn.benchmark = True
import pycuda.autoinit

    
import tensorrt as trt
import trt_utils as tu # trt utils
import pdb


TRT_LOGGER = trt.Logger()

def test_bs_1():
    #engine, context,  h_input, d_input, h_output, d_output, stream = onnx_2_tensorrt.main()
    onnx_file_path = 'rfb.onnx'
    fp16_mode = False#args.fp16
    int8_mode = False#args.int8
    batchsize =1 
    engine_file_path = "rfb_fp16_{}_int8_{}_bs_{}.trt".format(fp16_mode,int8_mode,batchsize)

    print("Building Engine")
  
    engine = tu.get_engine(batchsize,onnx_file_path,engine_file_path,fp16_mode=fp16_mode,int8_mode=int8_mode)
    #pdb.set_trace()
    context = engine.create_execution_context() 
    inputs, outputs, bindings, stream = tu.allocate_buffers(engine) # input, output: host # bindings 


    trt_outputs = []

    #img = np.random.randn(1,3,512,512) + 100
    #img = img.astype(np.uint8)
    global img
    x = img 
        
    inputs[0].host = x.reshape(-1)
    trt_outputs = tu.do_inference(context, bindings=bindings, inputs=inputs, outputs=outputs, stream=stream)
        # 
    locs, scores = tu.process_outputs(trt_outputs,batchsize, 6)  # The locs is corr
    #print(np.mean(locs))
    #print(np.mean(scores))
    pdb.set_trace()


def test_bs_2():
    onnx_file_path = 'rfb.onnx'
    fp16_mode = False#args.fp16
    int8_mode = False#args.int8
    batchsize = 2 
    engine_file_path = "rfb_fp16_{}_int8_{}_bs_{}.trt".format(fp16_mode,int8_mode,batchsize)

    print("Building Engine")
  
    engine = tu.get_engine(batchsize,onnx_file_path,engine_file_path,fp16_mode=fp16_mode,int8_mode=int8_mode)
    #pdb.set_trace()
    context = engine.create_execution_context() 
    inputs, outputs, bindings, stream = tu.allocate_buffers(engine) # input, output: host # bindings 

    trt_outputs = []

    #img = np.random.randn(1,3,512,512) + 100
    #img = img.astype(np.uint8)

    global img
    x = np.concatenate([img,img]) 
        
    inputs[0].host = x.reshape(-1)
    trt_outputs = tu.do_inference(context, bindings=bindings, inputs=inputs, outputs=outputs, stream=stream)
        # 
    locs, scores = tu.process_outputs(trt_outputs, batchsize, 6)  # The locs is c       
    #print(np.mean(locs))
    #print(np.mean(scores))
    pdb.set_trace()

if __name__ == '__main__':
    img = np.random.randn(1,3,512,512) + 100
    img = img.astype(np.uint8)
    test_bs_1()
    test_bs_2()
```

## Steps To Reproduce
python produce_bug.py

There are two test functions in the  produce_bug.py, 

In `test_bs_1()`, the code generate an engine whose maxBatchSize is 1, when I generate a random input img and set up a input `x = img` (batch size=1)   
 
![image](https://user-images.githubusercontent.com/37952703/96973842-04ab5580-154b-11eb-8939-68a292aa3b14.png)

the shape of locs is [1, 32756, 4] and the mean of `locs` is `-7.8930e+18` (non zero)

In `test_bs_2()`,  the code generate an engine whose maxBatchSize is 2, and the input becomes the concatenation of img that `x = np.concatenation([img, img])` (batch size =2)

 If the engine works correctly, the mean of `locs`.

![image](https://user-images.githubusercontent.com/37952703/96974111-57850d00-154b-11eb-8db2-a8e9fd70ba9c.png)

We can see from the screenshot above that the output `locs` has a lot of zeros values.
The first output `locs[0, :, :]` in the batch seems normal, but the second output `locs[1,: , : ]` in the batch has all zero values.
Besides, another output `boxes` have also a lot zero values when batch size > 1

Therefore, the TensorRT Engine does not work properly. 

 If I used TensorRT 6.0 to run the code,  `locs[1,: , : ]` will be normal (non zero).

This problem is very urgent for me. Please help me!
Appreciate any suggestions!

**Comment by CallmeZhangChenchen** [comment]
7.1里面加载onnx模型的方式变了，所以当batch size！=的时候会出现问题

**Comment by RizhaoCai** [comment]
> 7.1里面加载onnx模型的方式变了，所以当batch size！=的时候会出现问题

请问应该怎么解决呢，是在pytorch输出ONNX model的时候指定batch size还是在 parse ONNX model的时候，加一些什么参数。 求指导

**Comment by RizhaoCai** [comment]
Appreciate any help!

**Comment by NgTuong** [comment]
Hi, @RizhaoCai ,
i have same problem.
when run batch_size = 2. 
+ Inputs : array images define :inputs[0].host = images batch
+ Output: array 1 dimension. After reshape (batch, , ). results in first batch too correct, next batches is incorrect.
You have solution??

**Comment by RizhaoCai** [comment]
> Hi, @RizhaoCai ,
> i have same problem.
> when run batch_size = 2.
> 
> * Inputs : array images define :inputs[0].host = images batch
> * Output: array 1 dimension. After reshape (batch, , ). results in first batch too correct, next batches is incorrect.
>   You have solution??

Same issue. I have tried many methods. But all failed.

**Comment by ttyio** [comment]
Hello @RizhaoCai , thanks for reporting.
Now the onnx parser only support explicit batch dimensions since 7.x,  So we need export onnx with that batch dim as dynamic axis (-1) from framework, then call context.setBindingDimensions(...) during inference.

More details on this here: https://forums.developer.nvidia.com/t/tensorrt-7-onnx-models-with-variable-batch-size/115302/6?u=nves_r

**Comment by NgTuong** [comment]
Hi @ttyio 
Thanks for your answer.
You can explain for me here "batch dim as dynamic axis (-1) from framework".
I using python API to convert to onnx.
THanks

**Comment by ttyio** [comment]
Hello @NgTuong ,
see https://pytorch.org/docs/stable/onnx.html, we can specify `dynamic_axes` in the `torch.onnx.export` function.

**Comment by NgTuong** [comment]
@ttyio 
I fixed dynamic_axes in torch.onnx..export but result still fail.

**Comment by NgTuong** [comment]
![Uploading Screenshot from 2020-11-30 14-23-36.png…]()

**Comment by ttyio** [comment]
@NgTuong , did you use `createOptimizationProfile` in builder and `setBindingShape` in inference?

**Comment by NgTuong** [comment]
@ttyio 
I checked from model ONNX before check model trt. Results are failed.
If results in onnx step are failed, results in trt step too

**Comment by NgTuong** [comment]
![Screenshot from 2020-11-30 14-23-36](https://user-images.githubusercontent.com/42471216/100582136-5b126d80-331b-11eb-8c4a-888e6c424574.png)
Module convert to onnx. What's wrong?

**Comment by ttyio** [comment]
Hello @NgTuong , the export to onnx is correct. 
For how to use `createOptimizationProfile` in builder and `setBindingShape` in inference?
Could you check https://github.com/NVIDIA/TensorRT/tree/master/samples/opensource/trtexec

**Comment by NgTuong** [comment]
@ttyio  oki, thank you very much.
I will check it again and response for you.

**Comment by ttyio** [comment]
@NgTuong 
BTW, besides read the code, you can also run `trtexec` follow https://github.com/NVIDIA/TensorRT/tree/master/samples/opensource/trtexec#example-4-running-an-onnx-model-with-full-dimensions-and-dynamic-shapes
to first verify that the batchsize = 2 works when using `trtexec`.

**Comment by RizhaoCai** [comment]
> Hello @RizhaoCai , thanks for reporting.
> Now the onnx parser only support explicit batch dimensions since 7.x, So we need export onnx with that batch dim as dynamic axis (-1) from framework, then call context.setBindingDimensions(...) during inference.
> 
> More details on this here: https://forums.developer.nvidia.com/t/tensorrt-7-onnx-models-with-variable-batch-size/115302/6?u=nves_r

Hi @ttyio ,

Thanks for the suggestion. I have follow this and export onnx with (-1) for the batch axix.
When I build the tensorRT engine, I set a IOptimizationProfile. When executing the engine, I use `context.set_binding_shape`. But the problem still persist.

Pls see the image below, the output of `conext.get_binding_shape` is `(2,3,512,512)`, meaning that now tensorRT works with batch size = 2. However, the output `locs` still has the problem that the first element in the batch has normal results (values are not all zeros; you can see the top part of the output 
 of `locs`) but the other elements in the batch are all zero values (you can see the bottom part of the output of `locs`).
![image](https://user-images.githubusercontent.com/37952703/100952636-de1a0a80-354b-11eb-9236-670ba4435176.png).

Since I am using RFB-Net, which is an object detector, which would be complex. 
I will work out a toy example with the IOptimizationProfile to see if the problem still persists again.

**Comment by NgTuong** [comment]
Hi @RizhaoCai  Iuse HardNet, I convert with coco checkpoint, everything oki. But I test my checkpoint failed.
@ttyio , @RizhaoCai , I think that I train model with torch 0.4.0 but export using torch > =1.3. Does the version affects to export model?

**Comment by elevir** [comment]
TensorRT 7.2.2
The same problem, but not only with the batch axis. I need to infer a fully convolutional neural network and an image size with a batch size can change (but in my case it might change before any optimizations). If I remember correctly, before I could change an input node shape and everything worked fine, but now inferring always gets zero output. I've tried to load model with dynamic axes, but I need to add slice/reshape layers with which I've met errors such as 'negative axis' and 'out of bounds on axis' even in dynamic versions of these layers.

**Comment by RizhaoCai** [comment]
Hi @elevir @NgTuong 
Sorry, I got too busy for a long time and I did not do much on this issue. 
Yesterday, I manage to get the correct way to use the dynamic shape feature with a toy example. I will share the code here.

**Comment by RizhaoCai** [comment]
I close this issue because I have managed to achieve the expected results.

**Comment by dongyi-kim** [comment]
@ttyio I got a question about batch size. sorry for mention.
It's not a case using dynamic batch size.

My question is that, when my raw onnx model have batch size = 1, 
is there problem on creating trt enigine with batchsize=2?

**Comment by ttyio** [comment]
> @ttyio I got a question about batch size. sorry for mention. It's not a case using dynamic batch size.
> 
> My question is that, when my raw onnx model have batch size = 1, is there problem on creating trt enigine with batchsize=2?
> 
> only scenario with onnx.batch_size == trt.batch_size is acceptable?
> 
> I'm on the scenario, trt have wrong output.

Hi @dongyi-kim , there will be problem. we have to export onnx model which have batch size = -1 or 2 if you want creating trt engine with batch size 2. thanks!
